rust/capture/src/passes/invalid_surrogates.rs

```diff
@@ -0,0 +1,353 @@
+                        }
+                        Err(Some(c)) => {
+                            self.queue_str(REPLACEMENT);
+                            self.queue('\\');
+                            self.queue_str(REPLACEMENT);
+                            self.queue(c);
+                            return self.pop();
+                        }
+                    };
+                if HIGH_SURROGATE_RANGE.contains(&first_code_point)
+                    && LOW_SURROGATE_RANGE.contains(&second_code_point)
+                {
+                    // We have a valid pair of hex escapes, so we should push them.
+                    // TODO - there's way to do this that doesn't require the
+                    // allocation format! implies, but I'm not gonna work it out
+                    // right now - we expect this to be /extremely/ rare
+                    self.queue_str(&format!(
+                        "u{:04X}\\u{:04X}", // First backslash is already in the buffer due to last_seen
+                        first_code_point, second_code_point
+                    ));
+                } else {
+                    // We didn't get a valid pair, so we just drop the pair entirely
+                    self.queue_str(REPLACEMENT);
+                    self.queue('\\');
+                    self.queue_str(REPLACEMENT);
+                }
+            }
+            (LastSeen::Char | LastSeen::Escape, c) => {
+                // emit handles the transition between escape and char for us,
+                // so we just unconditionally emit here if the last thing we saw
+                // was a char, or the last thing we saw was an escape, AND the
+                // current char is not a 'u' (the case above)
+                self.queue(c);
+            }
+        }
+
+        // Because we swallow escape chars to avoid backtracking, we have to recurse
+        // here to handle the case where we just entered an escape squeuence
+        self.next()
+    }
+}
+
+// Collects 4 chars into a hex escape sequence, returning the first char that couldn't be part of
+// one, if one was found. If we run out of input, we return Result::Err(None)
+fn collect_escape_sequence(
+    buf: &mut String,
+    iter: &mut dyn Iterator<Item = char>,
+) -> Result<u16, Option<char>> {
+    buf.clear();
+    for _ in 0..HEX_ESCAPE_LENGTH {
+        let Some(c) = iter.next() else {
+            return Err(None);
+        };
+        // If this character couldn't be part of a hex escape sequence, we return it
+        if !c.is_ascii_hexdigit() {
+            return Err(Some(c));
+        }
+        buf.push(c);
+    }
+    // Unwrap safe due to the checking above
+    Ok(u16::from_str_radix(buf, 16).unwrap())
+}
+
+#[cfg(test)]
+mod test {
+
+    use crate::v0_request::RawEvent;
+
+    use super::InvalidSurrogatesPass;
+
+    const RAW_DATA: &str = include_str!("../../tests/invalid_surrogate.json");
+
+    #[test]
+    fn test() {
```

Comment:
unless this is common in rust we should give it a name 😊

Reply:
True, will do